### PR TITLE
`Development`: Add domain objects for standardized competencies

### PR DESCRIPTION
**Pull request by rstief**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
Adds the domain objects for [#8001](https://github.com/ls1intum/Artemis/issues/8001)

### Description
- Adds the tables, domain objects, repositories, services and resources for standardized competencies.
- Adds a `BaseCompetency` object from which all competencies inherit.

### Steps for Testing
Run locally and see that the migration works and Artemis starts up.
Test the following endpoints with Postman:
- POST api/admin/standardized-competencies/knowledge-areas (create a knowledge area)
- POST api/admin/standardized-competencies (create a standardized competency)
- GET api/standardized-competencies/:id (get a standardized competency)
- GET api/standardized-competencies/knowledge-areas/:id (get a knowledge area)

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [BaseCompetency.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6180/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.competency/BaseCompetency.html) | 100% | ✅ |
| [Competency.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6180/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.competency/Competency.html) | 79% | ✅  |
| [KnowledgeArea.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6180/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.competency/KnowledgeArea.html) | 100% | ✅ |
| [Source.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6180/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.competency/Source.html) | 92% | ✅  |
| [StandardizedCompetency.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6180/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.competency/StandardizedCompetency.html) | 89% | ✅  |
| [SourceRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6180/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/SourceRepository.html) | 100% | ✅ |
| [KnowledgeAreaRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6180/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository.competency/KnowledgeAreaRepository.html) | 100% | ✅ |
| [StandardizedCompetencyRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6180/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository.competency/StandardizedCompetencyRepository.html) | 100% | ✅ |
| [KnowledgeAreaService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6180/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.competency/KnowledgeAreaService.html) | 100% | ✅ |
| [StandardizedCompetencyService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6180/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.competency/StandardizedCompetencyService.html) | 100% | ✅ |
| [AdminStandardizedCompetencyResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6180/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.admin/AdminStandardizedCompetencyResource.html) | 100% | ✅ |
| [StandardizedCompetencyResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6180/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.competency/StandardizedCompetencyResource.html) | 100% | ✅ |

## Summary by CodeRabbit

- **New Features**
	- Introduced a service for efficiently managing knowledge areas, including creation and validation processes.
- **Tests**
	- Added comprehensive integration tests for standardized competencies and knowledge areas, ensuring robust authorization and error handling mechanisms.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/competency/StandardizedCompetencyService.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    private final SourceRepository sourceRepository;
+
+    public StandardizedCompetencyService(StandardizedCompetencyRepository standardizedCompetencyRepository, KnowledgeAreaRepository knowledgeAreaRepository,
+            SourceRepository sourceRepository) {
+        this.standardizedCompetencyRepository = standardizedCompetencyRepository;
+        this.knowledgeAreaRepository = knowledgeAreaRepository;
+        this.sourceRepository = sourceRepository;
+    }
+
+    public StandardizedCompetency createStandardizedCompetency(StandardizedCompetency competency) {
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/competency/StandardizedCompetencyService.java:39]: Missing JavaDoc

File: src/main/java/de/tum/in/www1/artemis/service/competency/StandardizedCompetencyService.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        competencyToCreate.setFirstVersion(competencyToCreate);
+        return standardizedCompetencyRepository.save(competencyToCreate);
+    }
+
+    /**
+     * Verifies that a standardized competency that should be created is valid or throws a BadRequestException
+     *
+     * @param competency the standardized competency to verify
+     */
+    public void standardizedCompetencyIsValidOrElseThrow(StandardizedCompetency competency) throws BadRequestException {
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/competency/StandardizedCompetencyService.java:67]: Should this be public? Any use of this method outside of this service would indicate that the logic should probably be part of it.

File: src/main/java/de/tum/in/www1/artemis/service/competency/KnowledgeAreaService.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+@Service
+public class KnowledgeAreaService {
+
+    private final KnowledgeAreaRepository knowledgeAreaRepository;
+
+    public KnowledgeAreaService(KnowledgeAreaRepository knowledgeAreaRepository) {
+        this.knowledgeAreaRepository = knowledgeAreaRepository;
+    }
+
+    public KnowledgeArea createKnowledgeArea(KnowledgeArea knowledgeArea) {
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/competency/KnowledgeAreaService.java:26]: Missing JavaDoc

File: src/main/java/de/tum/in/www1/artemis/service/competency/KnowledgeAreaService.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+        return createdKnowledgeArea;
+    }
+
+    /**
+     * Verifies that a knowledge area that should be created is valid or throws a BadRequestException
+     *
+     * @param knowledgeArea the knowledge area to verify
+     */
+    public void knowledgeAreaIsValidOrElseThrow(KnowledgeArea knowledgeArea) throws BadRequestException {
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/competency/KnowledgeAreaService.java:49]: Should this be public? Any use of this method outside of this service would indicate that the logic should probably be part of it.

File: src/test/java/de/tum/in/www1/artemis/competency/StandardizedCompetencyIntegrationTest.java
Comment (MaximilianAnzinger) [inline src/test/java/de/tum/in/www1/artemis/competency/StandardizedCompetencyIntegrationTest.java:1]: Depending on today's dev-meeting let's utilize nested tests to improve readability.

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/repository/competency/KnowledgeAreaRepository.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    @NotNull
+    default KnowledgeArea findByIdElseThrow(long knowledgeAreaId) throws EntityNotFoundException {
+        return findById(knowledgeAreaId).orElseThrow(() -> new EntityNotFoundException("KnowledgeArea", knowledgeAreaId));
+    }
+
+    @Query("""
+            SELECT knowledgeArea FROM KnowledgeArea knowledgeArea
+                LEFT JOIN FETCH knowledgeArea.children
+                LEFT JOIN FETCH knowledgeArea.competencies
+            WHERE knowledgeArea.id = :knowledgeAreaId
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/repository/competency/KnowledgeAreaRepository.java:34]:
```suggestion
            SELECT knowledgeArea 
            FROM KnowledgeArea knowledgeArea
                LEFT JOIN FETCH knowledgeArea.children
                LEFT JOIN FETCH knowledgeArea.competencies
            WHERE knowledgeArea.id = :knowledgeAreaId
```

File: src/main/java/de/tum/in/www1/artemis/service/competency/KnowledgeAreaService.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            parent = knowledgeAreaRepository.findByIdElseThrow(parent.getId());
+        }
+
+        var knowledgeAreaToCreate = new KnowledgeArea(knowledgeArea.getTitle(), knowledgeArea.getDescription());
+        knowledgeAreaToCreate.setParent(parent);
+        var createdKnowledgeArea = knowledgeAreaRepository.save(knowledgeAreaToCreate);
+        // set the parent as it is lazily loaded
+        createdKnowledgeArea.setParent(parent);
+
+        return createdKnowledgeArea;
```
Comment (rstief) [inline src/main/java/de/tum/in/www1/artemis/service/competency/KnowledgeAreaService.java:47]:
```suggestion
        return knowledgeAreaRepository.save(knowledgeAreaToCreate);
```

**Review by reschandreas — APPROVED** [review]
code lgtm!

**Review by laurenzfb — DISMISSED** [review]
tested using postman on ts1. endpoints work

**Review by mateusmm01 — APPROVED** [review]
Demonstrated in testing session. Looks good!

**Review by reschandreas — APPROVED** [review]

**Review by milljoniaer — APPROVED** [review]
Code looks good 👍

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]
Please check the tests again.

**Review by MaximilianAnzinger — DISMISSED** [review]
Only missing import, no reapprovals required 👍

**Comment by rstief** [comment]
Blocked by the spring boot upgrade.

**Review by MaximilianAnzinger — APPROVED** [review]

**Review by milljoniaer — APPROVED** [review]
Code looks good 👍

**Review by reschandreas — APPROVED** [review]

**Review by JohannesWt — DISMISSED** [review]
Tested in virtual test session on test server 2 with @rstief . Everything works 👍

**Review by MaximilianAnzinger — APPROVED** [review]
LGTM

**Review by MaximilianAnzinger — APPROVED** [review]

**Pull request merged**